src/snowflake/snowpark/_internal/analyzer/select_statement.py

```diff
@@ -382,7 +382,7 @@ def column_states(self, value: ColumnStateDict):
 
     @property
     @abstractmethod
-    def referenced_ctes(self) -> Set[WithQueryBlock]:
+    def referenced_ctes(self) -> Dict[WithQueryBlock, int]:
         """Return the set of ctes referenced by the whole selectable subtree, includes its-self and children"""
```

Comment:
we will need to update the comments for referenced_ctes to indicate the changes, where it also records the number of times each CTE is referred

Reply:
fixed